### PR TITLE
refactor: use `Duration` instead of `u32`

### PR DESCRIPTION
**Pull request by CrabNejonas**
Use the proper `Duration` instead of `u32` for time values such as ttl.

This leads to less error prone and more readable code since plain `u32` values carry no information about unit. Using `Duration ` ensures time values are always correctly given as seconds.

### CODE REVIEW

**Comment by partim** [comment]
Thank you for the PR!

I agree that we should use a dedicated type for TTL values rather than just the raw `u32`. However, I think because `Duration` provides sub-second precision it is the wrong choice here. I think it would be better to have a newtype `Ttl` (it could live in `base::record`) around the `u32` that provides explicit conversions to and from `Duration`.

Would you be interested to adjust your PR accordingly – otherwise I can also do the changes.

**Comment by CrabNejonas** [comment]
Good point, using `Duration` isn't really applicable here and it also doesn't expose information commonly used when working with durations in DNS (e.g. ttl in minutes or hours). I added a custom `Ttl` type that wraps a `u32` and mirrors the methods exposed by `Duration`.

Only issue I have with the change is the naming, `Ttl` feels like a weird name for this. I was thinking maybe `LowPrecisionDuration` or `LongDuration` would be more appropriate? But also way more verbose ofc

**Comment by partim** [comment]
I agree `Ttl` isn’t perfect, but then again it is super short which is nice, and basically everyone working with DNS knows what it implies. So I think we should just go with it.

**Review by partim — CHANGES_REQUESTED** [review]

**Comment by partim** [comment]
There’s a weird quirk in RFC 1034 (or 1035) where is mistakenly declared TTL at one point to be a signed number. RFC 2181 later made it officially unsigned but said that any value with the most significant bit said should be treated as zero. Years later, RFC 8767 suggested that that was a bad idea and made the TTL properly `u32` but with the caveat that values SHOULD be capped somewhere in the days or weeks, suggesting 7 days (604,800 seconds) as the cap.

Which is a long winded way to suggest that `Ttl` should probably have a `cap` method that simply caps it at those 7 days – perhaps also a constant CAP with that value.

**Comment by CrabNejonas** [comment]
> There’s a weird quirk in RFC 1034 (or 1035) where is mistakenly declared TTL at one point to be a signed number. RFC 2181 later made it officially unsigned but said that any value with the most significant bit said should be treated as zero. Years later, RFC 8767 suggested that that was a bad idea and made the TTL properly `u32` but with the caveat that values SHOULD be capped somewhere in the days or weeks, suggesting 7 days (604,800 seconds) as the cap.
> 
> Which is a long winded way to suggest that `Ttl` should probably have a `cap` method that simply caps it at those 7 days – perhaps also a constant CAP with that value.

That's incredibly weird hahaha, I didn't know. But yeah that makes sense 👍

File: src/rdata/rfc1035.rs [hunk 1]
```diff
@@ -1257,24 +1258,24 @@ impl<N> Soa<N> {
         self.serial
     }
 
-    /// The time interval in seconds before the zone should be refreshed.
-    pub fn refresh(&self) -> u32 {
-        self.refresh
+    /// The time interval before the zone should be refreshed.
+    pub fn refresh(&self) -> &Ttl {
```
[the remark this answers is not in the document]
Comment (CrabNejonas) [reply src/rdata/rfc1035.rs:1262]: oops yep

File: src/base/record.rs [hunk 2]
```diff
@@ -532,8 +567,8 @@ impl<Name> RecordHeader<Name> {
     }
 
     /// Returns the TTL of the record.
-    pub fn ttl(&self) -> u32 {
-        self.ttl
+    pub fn ttl(&self) -> &Ttl {
```
[the remark this answers is not in the document]
Comment (CrabNejonas) [reply src/base/record.rs]: done

File: src/base/wire.rs [hunk 3]
```diff
@@ -148,6 +149,17 @@ impl Compose for Ipv6Addr {
     }
 }
 
+impl Compose for Ttl {
```
[the remark this answers is not in the document]
Comment (CrabNejonas) [reply src/base/wire.rs]: done

File: src/base/wire.rs [hunk 4]
```diff
@@ -222,6 +246,12 @@ impl<'a, Octs: AsRef<[u8]> + ?Sized> Parse<'a, Octs> for Ipv6Addr {
     }
 }
 
+impl<'a, Octs: AsRef<[u8]> + ?Sized> Parse<'a, Octs> for Ttl {
```
[the remark this answers is not in the document]
Comment (CrabNejonas) [reply src/base/wire.rs]: done

**Comment by CrabNejonas** [comment]
Alright, I added doc comments + examples to all methods mirroring std::time::Duration. A few things of note:

`Ttl` doesn't implement `From<Duration>` as the conversion is lossy (Duration can represent sub-second values and stores seconds in `u64`) so conversion is exposed as `Ttl::from_duration_lossy` it does however implement `Into<Duration>` because that conversion can be done without an issue.

In examples I used `core::time::Duration` because the rest of the crate has to work in no_std, but maybe this is confusing?

**Comment by partim** [comment]
I think having the examples require the _std_ feature and then use that us okay. Probably less confusing, indeed.

**Comment by CrabNejonas** [comment]
> I think having the examples require the _std_ feature and then use that us okay. Probably less confusing, indeed.

yep, done!

**Review by partim — APPROVED** [review]
Again, thank you for this very nice PR! Also thank you for the quick turn-around – this allows us to include it in the soon-to-be-released 0.8.0.

**Pull request merged**